forc-pkg/src/pkg.rs

```diff
@@ -750,20 +773,79 @@ pub fn compilation_order(graph: &Graph) -> Result<Vec<NodeIx>> {
+/// is actually possible to build.
+///
+///
+/// There might be some differences between the graph that is described by the lock file and
+/// the graph that can actually possible to build. This is happening in the case of a dependency
+/// (described by path) removal after the lock file is generated. Since dependency is removed
+/// graph_to_path_map cannot find the required path for the dependency entry in the lock file.
+/// This requires us to remove those removed nodes from the graph so that we can have a BuildPlan
+/// generated from lock file. After the BuildPlan is generated it can be used to apply diffs by
+/// apply_pkg_diff().
+fn apply_diff_after_lock(
+    deps_to_remove: HashSet<NodeIx>,
+    graph: &mut Graph,
+    path_map: &PathMap,
+    sway_git_tag: &str,
+) -> Result<()> {
+    for dep in deps_to_remove {
+        let dep_name = &graph[dep].name.clone();
+
+        let (parent_node, _) = graph
+            .edges_directed(dep, Direction::Incoming)
+            .next()
+            .map(|edge| (edge.source(), edge.weight().clone()))
+            .ok_or_else(|| anyhow!("more than one root package detected in graph"))?;
+
+        let parent = &graph[parent_node].clone();
+        // Construct the path relative to the parent's path.
+        let parent_path = &path_map[&parent.id()];
+        let parent_manifest = ManifestFile::from_dir(parent_path, sway_git_tag)?;
+
+        // Check if parent's manifest file includes this dependency.
+        // If not this is a removed dependency and we should remove the removed dependency from the graph.
+        if parent_manifest.dep(dep_name).is_none() {
+            // We should be connecting the parent of the removed dependency to each child of removed node.
+            // Consider the following scenerio: a -> b -> c
+            // If you remove b, the edges between b -> c will be unvalidated so while processing c, you cannot find the parent.
+            // Instead you need to fetch all children of b and add an edge between those children and a before removing.
+
+            // Get children of the node to delete
+            let child_nodes: Vec<NodeIx> = graph
+                .edges_directed(dep, Direction::Outgoing)
+                .map(|edges| (edges.target()))
+                .collect();
+            // Add an edge between the node that is going to be removed's child nodes and parent of the same node.
+            for child in child_nodes {
+                // Check if there is already an edge if there is not add it
+                if !graph.edges(parent_node).any(|edge| edge.target() == child) {
+                    graph.add_edge(parent_node, child, graph[child].name.clone());
+                }
+            }
+            println_red(&format!("  Removing {}", dep_name));
+            // Remove the deleted node
+            graph.remove_node(dep);
+        }
+    }
+    Ok(())
+}
```

Comment:
Is it possible to re-use `remove_deps` in this implementation?

Reply:
This got me thinking and realized that this approach was wrong for handling removed packages after the lock file is generated. 

What I was doing was while visiting in the parent-to-child direction I was looking for the current node's parent's manifest file for the same reason if I cannot find the dependency,  I was deleting the current node and connecting the current node's child nodes to the current node's parent. 

**An example scenario where this seems wrong to me:**

If `a` -> (depends on) `b` -> (depends on) `c`, and If I remove `b`, `c` was getting connected to `a` and this is not necessarily correct. Imagine a package depending on `std` and we are removing `std` from the manifest file (in the case of an explicit declaration) after the lock file is generated, this approach would result in `core` to be getting connected to the project_node.

We should be visiting in the child-to-parent direction and checking if the parent of that node's manifest file does not contain a dependency declaration for the current node, if that is the case current node is a removed dependency that should be removed from the dependency graph generated from the lock file.

I think it seems not easily possible to re-use `remove_deps` since we should be visiting nodes in the child-to-parent manner, which is the opposite direction of the `removed_deps` and also, different from the `removed_deps` we need to be checking parent's manifest file. So the operation seems different enough to create the need for a new implementation but I am open to any suggestion as it would be looking better if we can re-use it somehow.

Reply:
> we need to be checking parent's manifest file

Why do we need to check the parent's manifest file here? It seems to me that we have already done this check when we collected this set of `deps_to_remove`?

Reply:
Yep... I confused myself unnecessarily since we already collected the required set of deps_to_remove we should not check it again.